provider/oci/images_test.go

```diff
@@ -0,0 +1,393 @@
+	defer ctrl.Finish()
+
+	listImageRequest, listImageResponse := makeListImageRequestResponse(
+		i.testCompartment, []map[string]string{
+			map[string]string{
+				"id":      "fakeUbuntu1",
+				"os":      "Canonical Ubuntu",
+				"name":    "Canonical-Ubuntu-14.04-2018.01.11-0",
+				"version": "14.04",
+			},
+			map[string]string{
+				"id":      "fake image id for bad image",
+				"os":      "CentOS",
+				"name":    "BadlyFormatedDisplayName_IshouldBeIgnored",
+				"version": "7",
+			},
+		})
+
+	shapesRequestUbuntu, shapesResponseUbuntu := makeShapesRequestResponse(
+		i.testCompartment, "fakeUbuntu1", []string{"VM.Standard2.1", "VM.Standard1.2"})
+
+	shapesRequestBadImage, shapesResponseBadImage := makeShapesRequestResponse(
+		i.testCompartment, "fake image id for bad image", []string{"VM.Standard1.2"})
+
+	gomock.InOrder(
+		compute.EXPECT().ListImages(context.Background(), listImageRequest).Return(listImageResponse, nil),
+		compute.EXPECT().ListShapes(context.Background(), shapesRequestUbuntu).Return(shapesResponseUbuntu, nil),
+		compute.EXPECT().ListShapes(context.Background(), shapesRequestBadImage).Return(shapesResponseBadImage, nil),
+	)
+
+	imgCache, err := oci.RefreshImageCache(compute, &i.testCompartment)
+	c.Assert(err, gc.IsNil)
+	c.Assert(imgCache, gc.NotNil)
+	c.Check(len(imgCache.ImageMap()), gc.Equals, 1)
+	imageMap := imgCache.ImageMap()
+
+	c.Check(imageMap["trusty"][0].Id, gc.Equals, "fakeUbuntu1")
+}
+
+func (i *imagesSuite) TestRefreshImageCacheWithAPIError(c *gc.C) {
+	ctrl := gomock.NewController(c)
+	compute := ocitesting.NewMockOCIComputeClient(ctrl)
+	defer ctrl.Finish()
+
+	compute.EXPECT().ListImages(gomock.Any(), gomock.Any()).Return(ociCore.ListImagesResponse{}, errors.Errorf("awww snap!")).Times(1)
+	_, err := oci.RefreshImageCache(compute, &i.testCompartment)
+	c.Assert(err, gc.ErrorMatches, "awww snap!")
+}
+
+func (i *imagesSuite) TestImageMetadataFromCache(c *gc.C) {
+	image := ociCore.Image{
+		CompartmentId:          &i.testCompartment,
+		Id:                     &i.testImageID,
+		OperatingSystem:        makeStringPointer("Canonical Ubuntu"),
+		OperatingSystemVersion: makeStringPointer("14.04"),
+		DisplayName:            makeStringPointer("Canonical-Ubuntu-14.04-2018.01.11-0"),
+	}
+
+	imgType, err := oci.NewInstanceImage(image, &i.testCompartment)
+	c.Assert(err, gc.IsNil)
+
+	cache := &oci.ImageCache{}
+	images := map[string][]oci.InstanceImage{
+		"trusty": []oci.InstanceImage{
+			imgType,
+		},
+	}
+	cache.SetImages(images)
+	metadata := cache.ImageMetadata("trusty", "")
+	c.Assert(len(metadata), gc.Equals, 1)
+	// generic images default to ImageTypeVM
+	c.Assert(metadata[0].VirtType, gc.Equals, string(oci.ImageTypeVM))
+
+	// explicitly set ImageTypeBM on generic images
+	metadata = cache.ImageMetadata("trusty", string(oci.ImageTypeBM))
+	c.Assert(len(metadata), gc.Equals, 1)
+	c.Assert(metadata[0].VirtType, gc.Equals, string(oci.ImageTypeBM))
+}
```

Comment:
I think we're missing some of the coverage around how we handle the DefaultImageType and values that are in the metadata.
Namely, the fact that if the type *isn't* Generic then the defaultVirtType has no effect.
I think adding a comment before these tests stating:
There are 4 cases:
1) Generic Image, no default type, you get ImageTypeVM
2) Generic Image, default type, you get default type
3) Non-generic image, no default type you get specific image type
4) Non-generic image, default type you still get specific image type

I think you tested the first two cases, but we need to cover the third as well.
Can you also update the DocString for ImageMetadata to clarify that as well?
Thanks

Reply:
done